libs/cli/src/main/java/org/elasticsearch/cli/internal/JsonPrintWriter.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.cli.internal;
+
+import org.elasticsearch.core.SuppressForbidden;
+
+import java.io.OutputStream;
+import java.io.PrintWriter;
+import java.time.Clock;
+import java.time.Instant;
+import java.time.format.DateTimeFormatter;
+import java.time.format.DateTimeFormatterBuilder;
+import java.util.Locale;
+
+/**
+ * {@link PrintWriter} formatting lines as JSON using a subset of fields written by ECSJsonLayout.
+ * Anything that looks like a JSON object is immediately passed through and written as is.
+ */
+public class JsonPrintWriter extends PrintWriter {
+    private static final int INITIAL_BUFFER_SIZE = 256;
+    private static final int MAX_BUFFER_SIZE = 4 * INITIAL_BUFFER_SIZE;
+    private static final DateTimeFormatter ISO_INSTANT_FORMATTER = new DateTimeFormatterBuilder().appendInstant(3).toFormatter(Locale.ROOT);
+
+    private final StringBuilder buffer = new StringBuilder(INITIAL_BUFFER_SIZE);
+    private final Clock clock;
+
+    public JsonPrintWriter(OutputStream out, boolean autoFlush) {
+        this(out, autoFlush, Clock.systemUTC());
+    }
+
+    @SuppressForbidden(reason = "Override PrintWriter to emit Json")
+    protected JsonPrintWriter(OutputStream out, boolean autoFlush, Clock clock) {
+        super(out, autoFlush);
+        this.clock = clock;
+    }
+
+    @Override
+    public void write(String msg) {
+        if (isJsonObject(msg)) {
+            super.write(msg);
+        } else {
+            synchronized (buffer) {
+                buffer.append(msg);
```

Comment:
Shouldn't this be calling toJson?

Reply:
Not in the current implementation, currently there's a 2nd buffer for building json. Admittedly, that could be optimized to just use a single buffer and jsonize on the fly, but I haven't done that yet. Let's agree how to proceed below first.